content/blog/2020-03-dplyr-1-0-0-and-vctrs/index.Rmarkdown

```diff
@@ -0,0 +1,149 @@
+---
+title: dplyr 1.0.0 and vctrs
+author: Hadley Wickham
+date: '2020-03-25'
+slug: dplyr-1-0-0-and-vctrs
+photo:
+  url: https://unsplash.com/photos/IstXvxHGoA4
+  author: 35mm
+categories:
+  - package
+tags:
+  - dplyr
+---
+
+[dplyr 1.0.0 is coming soon](https://www.tidyverse.org/blog/2020/03/dplyr-1-0-0-is-coming-soon/), and we've already shown you how [`summarise()` is growing](https://www.tidyverse.org/blog/2020/03/dplyr-1-0-0-summarise/). Today, I wanted to talk a little bit how dplyr 1.0.0 now uses the  [vctrs](http://vctrs.r-lib.org/) package. This post explains why vctrs is so important, why we can't just copy what base R does, how to interpret some of new error messages that you'll see, and some of the major changes since the last version.
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
+```
+
+## Combining vectors
+
+The heart of the reason we're using vctrs is the need to combine vectors together. You're already familiar with one base R tool for combining vectors, `c()`:
+
+```{r}
+c(1, 2, 3)
+c("a", "b", "c")
+```
+
+But combining vectors comes up in many places in the tidyverse, e.g.:
+
+* `dplyr::mutate()` and `dplyr::summarise()` have to combine the results 
+  from each group.
+* `dplyr::bind_rows()` has to combine columns from different data frames.
+* `dplyr::full_join()`  has to combine the keys from the `x` and `y` data
+  frames.
+* `tidyr::pivot_longer()` has to combine multiple columns into one.
+
+Our goal is to unify the code that underlies all these various functions so that there's one consistent, principled approach. We've already made the change in [tidyr](https://www.tidyverse.org/blog/2019/09/tidyr-1-0-0/), and now it's dplyr's turn. 
+
+## Base R behaviour
+
+You might wonder why we can't just copy the behaviour of `c()`. Unfortunately `c()` has some major downsides:
+
+*   It converts factors to their underlying integer levels:
+    
+    ```{r}
+    c(factor("x"), factor("y"))
+    ```
+  
+*   It doesn't yield sensible results when combining dates and times:
+
+    ```{r}
+    today <- as.Date("2020-03-24")
+    now <- as.POSIXct("2020-03-24 10:34")
+    c(today, now)
+    c(now, today)
+    ```
+
+Additionally, `c()` isn't the only way that base R combines together. `rbind()` and `unlist()` can also be used to perform a similar job, but return different results. This is not to claim that the tidyverse has been any better in the past - we have used a variety of ad hoc method, undoubtedly using well more than three different approaches. 
+
+So decided to come up with our own equivalent of `c()` that we call `vectrs::vec_c()`. `vec_c()`'s behaviour is governed by two main principles:
+
+* Symmetry: `vec_c(x, y)` should return a type as similar as possible to 
+  `vec_c(y, x)`.
+
+* Decrease constraints: this is a bit harder to explain, but if you combine
```

Comment:
* Preservation of information: If the type of one object can represent more information than the type of another object, the richer type wins.

* Consistency: If the types do not represent the same kind of information (dates, numbers, strings, ...), this is an incompatible type error.